File: 2020/21/2020-21.jl
tds = """
mxmxvkd kfcds sqjhc nhms (contains dairy, fish)
trh fvjkl sbzzf mxmxvkd (contains dairy)
sqjhc fvjkl (contains soy)
sqjhc mxmxvkd sbzzf (contains fish)"""

function read_data(inp)
    r = []
    for line in split(strip(inp), "\n")
        line = line[1:end-1]  # final )
        line = replace(line, "," => "")
        ings,algs = split(line, "(contains ")
        push!(r, (Set(split(ings)), Set(split(algs))))
    end
    r
end

function possible_assignments(rules)
    all_algs = reduce(union, (algs for (ings,algs) in rules))
    can_be = Dict{String,Set{String}}()
    for alg in all_algs
        can_be[alg] = reduce(intersect, (is for (is,as) in rules if alg in as))
    end
    can_be
end

function part1(rules)
    can_be = possible_assignments(rules)
    all_used_ings = reduce(union, values(can_be))
    sm = 0
    for (ings,algs) in rules
        sm += count(i->!(i in all_used_ings), ings)
    end
    sm
end

test_rules = read_data(tds)
@assert part1(test_rules) == 5

rules = read_data(read("input.txt", String))
println("PART1: ", part1(rules))

function part2(rules)
    can_be = possible_assignments(rules)
    println(can_be)
end

println("PART2: ", part2(test_rules))

File: 2020/21/2020-21.py

tds = """
mxmxvkd kfcds sqjhc nhms (contains dairy, fish)
trh fvjkl sbzzf mxmxvkd (contains dairy)
sqjhc fvjkl (contains soy)
sqjhc mxmxvkd sbzzf (contains fish)"""

def read_data(inp):
    rs = []
    for line in inp.strip().splitlines():
        line = line.rstrip(")")
        line = line.replace(",","")
        ings_s, algs_s = line.split("(contains ")
        rs.append((set(ings_s.split()), set(algs_s.split())))
    return rs

def possible_assignments(rules):
    all_algs = set.union(*[algs for (ings,algs) in rules])
    can_be = {}
    for alg in all_algs:
        can_be[alg] = set.intersection(*[ings for (ings,algs) in rules if alg in algs])
    return can_be

def part1(rules):
    can_be = possible_assignments(rules)
    all_used_ings = set.union(*can_be.values())
    sm = 0
    for (ings,algs) in rules:
        sm += sum(1 for i in ings if i not in all_used_ings)
    return sm

test_rules = read_data(tds)
assert part1(test_rules) == 5

inp = open("input.txt").read()
rules = read_data(inp)
print("PART1:", part1(rules))

